### DISCUSSION TITLE
Need little help with AnimatedGifPanel + Button

### DISCUSSION THREAD

**Discussion by GrimorJander** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#297] [Ideas]
First of all I want to thank you for this incredible library that has allowed me to make the most of my matrix rgb panels.

I am trying to implement the inner button of the Wroom32 dev board, so that it goes from one gif to the next, in a similar way as it does in the ChainedPanelsAuroraDemo example.
I need that when boot, plays first gif in loop, until i press button, that plays second gif forever, press button next gif in loop and so on.
That way I can play a single gif, but choosing it from several pre-loaded.
So, I'm still a novice and I can only cut and paste code, but I still have a hard time understanding the bugs.
Here´s the code, that only plays first on a loop with button not working properly:

`// Example sketch which shows how to display a 64x32 animated GIF image stored in FLASH memory
// on a 64x32 LED matrix
//
// Credits: https://github.com/bitbank2/AnimatedGIF/tree/master/examples/ESP32_LEDMatrix_I2S
// 

#define FILESYSTEM SPIFFS
#include <SPIFFS.h>
#include <AnimatedGIF.h>
#include <ESP32-HUB75-MatrixPanel-I2S-DMA.h>

// ----------------------------

#define PANEL_RES_X 64      // Number of pixels wide of each INDIVIDUAL panel module. 
#define PANEL_RES_Y 32     // Number of pixels tall of each INDIVIDUAL panel module.
#define PANEL_CHAIN 1      // Total number of panels chained one to another

#define BTN_PIN 0 // Gif advance. Using EPS32 Boot button.
int buttonState;             // the current reading from the input pin
int lastButtonState = LOW;   // the previous reading from the input pin
unsigned long lastDebounceTime = 0;  // the last time the output pin was toggled
unsigned long debounceDelay = 50;    // the debounce time; increase if the output flickers

 
//MatrixPanel_I2S_DMA dma_display;
MatrixPanel_I2S_DMA *dma_display = nullptr;

uint16_t myBLACK = dma_display->color565(0, 0, 0);
uint16_t myWHITE = dma_display->color565(255, 255, 255);
uint16_t myRED = dma_display->color565(255, 0, 0);
uint16_t myGREEN = dma_display->color565(0, 255, 0);
uint16_t myBLUE = dma_display->color565(0, 0, 255);

AnimatedGIF gif;
File f;
int x_offset, y_offset;

// Draw a line of image directly on the LED Matrix
void GIFDraw(GIFDRAW *pDraw)
{
    uint8_t *s;
    uint16_t *d, *usPalette, usTemp[320];
    int x, y, iWidth;

  iWidth = pDraw->iWidth;
  if (iWidth > MATRIX_WIDTH)
      iWidth = MATRIX_WIDTH;

    usPalette = pDraw->pPalette;
    y = pDraw->iY + pDraw->y; // current line
    
    s = pDraw->pPixels;
    if (pDraw->ucDisposalMethod == 2) // restore to background color
    {
      for (x=0; x<iWidth; x++)
      {
        if (s[x] == pDraw->ucTransparent)
           s[x] = pDraw->ucBackground;
      }
      pDraw->ucHasTransparency = 0;
    }
    // Apply the new pixels to the main image
    if (pDraw->ucHasTransparency) // if transparency used
    {
      uint8_t *pEnd, c, ucTransparent = pDraw->ucTransparent;
      int x, iCount;
      pEnd = s + pDraw->iWidth;
      x = 0;
      iCount = 0; // count non-transparent pixels
      while(x < pDraw->iWidth)
      {
        c = ucTransparent-1;
        d = usTemp;
        while (c != ucTransparent && s < pEnd)
        {
          c = *s++;
          if (c == ucTransparent) // done, stop
          {
            s--; // back up to treat it like transparent
          }
          else // opaque
          {
             *d++ = usPalette[c];
             iCount++;
          }
        } // while looking for opaque pixels
        if (iCount) // any opaque pixels?
        {
          for(int xOffset = 0; xOffset < iCount; xOffset++ ){
            dma_display->drawPixel(x + xOffset, y, usTemp[xOffset]); // 565 Color Format
          }
          x += iCount;
          iCount = 0;
        }
        // no, look for a run of transparent pixels
        c = ucTransparent;
        while (c == ucTransparent && s < pEnd)
        {
          c = *s++;
          if (c == ucTransparent)
             iCount++;
          else
             s--; 
        }
        if (iCount)
        {
          x += iCount; // skip these
          iCount = 0;
        }
      }
    }
    else // does not have transparency
    {
      s = pDraw->pPixels;
      // Translate the 8-bit pixels through the RGB565 palette (already byte reversed)
      for (x=0; x<pDraw->iWidth; x++)
      {
        dma_display->drawPixel(x, y, usPalette[*s++]); // color 565
      }
    }
} /* GIFDraw() */

void * GIFOpenFile(const char *fname, int32_t *pSize)
{
  Serial.print("Playing gif: ");
  Serial.println(fname);
  f = FILESYSTEM.open(fname);
  if (f)
  {
    *pSize = f.size();
    return (void *)&f;
  }
  return NULL;
} /* GIFOpenFile() */

void GIFCloseFile(void *pHandle)
{
  File *f = static_cast<File *>(pHandle);
  if (f != NULL)
     f->close();
} /* GIFCloseFile() */

int32_t GIFReadFile(GIFFILE *pFile, uint8_t *pBuf, int32_t iLen)
{
    int32_t iBytesRead;
    iBytesRead = iLen;
    File *f = static_cast<File *>(pFile->fHandle);
    // Note: If you read a file all the way to the last byte, seek() stops working
    if ((pFile->iSize - pFile->iPos) < iLen)
       iBytesRead = pFile->iSize - pFile->iPos - 1; // <-- ugly work-around
    if (iBytesRead <= 0)
       return 0;
    iBytesRead = (int32_t)f->read(pBuf, iBytesRead);
    pFile->iPos = f->position();
    return iBytesRead;
} /* GIFReadFile() */

int32_t GIFSeekFile(GIFFILE *pFile, int32_t iPosition)
{ 
  int i = micros();
  File *f = static_cast<File *>(pFile->fHandle);
  f->seek(iPosition);
  pFile->iPos = (int32_t)f->position();
  i = micros() - i;
//  Serial.printf("Seek time = %d us\n", i);
  return pFile->iPos;
} /* GIFSeekFile() */

unsigned long start_tick = 0;

void ShowGIF(char *name)
{
  start_tick = millis();
   
  if (gif.open(name, GIFOpenFile, GIFCloseFile, GIFReadFile, GIFSeekFile, GIFDraw))
  {
    x_offset = (MATRIX_WIDTH - gif.getCanvasWidth())/2;
    if (x_offset < 0) x_offset = 0;
    y_offset = (MATRIX_HEIGHT - gif.getCanvasHeight())/2;
    if (y_offset < 0) y_offset = 0;
    Serial.printf("Successfully opened GIF; Canvas size = %d x %d\n", gif.getCanvasWidth(), gif.getCanvasHeight());
    Serial.flush();
    while (gif.playFrame(true, NULL))
    {      
      if ( (millis() - start_tick) > 8000) { // we'll get bored after about 8 seconds of the same looping gif
        break;
      }
    }
    gif.close();
  }

} /* ShowGIF() */

/************************* Arduino Sketch Setup and Loop() *******************************/
void setup() {

//

  HUB75_I2S_CFG mxconfig(
    PANEL_RES_X,   // module width
    PANEL_RES_Y,   // module height
    PANEL_CHAIN    // Chain length
  );

  mxconfig.gpio.e = 18;
  mxconfig.clkphase = false;
  mxconfig.driver = HUB75_I2S_CFG::FM6126A;

  // Display Setup
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);
  dma_display->begin();
  dma_display->setBrightness8(90); //0-255
  dma_display->clearScreen();
  dma_display->fillScreen(myWHITE);
  
//

  Serial.begin(115200);
  Serial.println("Starting AnimatedGIFs Sketch");

  // Start filesystem
  Serial.println(" * Loading SPIFFS");
  if(!SPIFFS.begin()){
        Serial.println("SPIFFS Mount Failed");
  }
  
  dma_display->begin();
  
  /* all other pixel drawing functions can only be called after .begin() */
  dma_display->fillScreen(dma_display->color565(0, 0, 0));
  gif.begin(LITTLE_ENDIAN_PIXELS);

}

String gifDir = "/gifs"; // play all GIFs in this directory on the SD card
char filePath[256] = { 0 };
File root, gifFile;

void loop() 
{
   while (1) // run forever
   {
      
      root = FILESYSTEM.open(gifDir);
      if (root)
      {
           gifFile = root.openNextFile();
            while (gifFile)
            {
              if (!gifFile.isDirectory()) // play it
              {
                
                // C-strings... urghh...                
                memset(filePath, 0x0, sizeof(filePath));                
                strcpy(filePath, gifFile.path());
                
                // Show it.
                ShowGIF(filePath);
               
              }

// Boot button Gif advance with debounce
    int reading = digitalRead(BTN_PIN);
    if (reading != lastButtonState) {
      lastDebounceTime = millis();
    }
    
    if ((millis() - lastDebounceTime) > debounceDelay) {
      if (reading != buttonState) {
        buttonState = reading;
        if (buttonState == LOW) {
          Serial.println("NEXT GIF...");
           gifFile.close();
              gifFile = root.openNextFile();
            }
         root.close();
      } // root
        }
      }
    }
    
    // end button debounce
              
             
      
      delay(1000); // pause before restarting
      
   } // while
}`

Any help will be apreciated.

**Comment by board707** [comment]
Your cycle 

while (gifFile) [ ...]

is endless. Since the first file in dir is viewed, the condition of while() still true and the program never leave the cycle and never changing the image and not read the button.

As a fix try to replace while() to the if() and place button reading code inside the loop

**Reply by GrimorJander** [reply]
Thanks for your reply. I believe that button routine was already inside the loop.

I try to change a little as you said, but button doesn´t play next file:

`void loop() 
{
   while (1) // run forever
   {
      
      root = FILESYSTEM.open(gifDir);
      if (root)
      {
           gifFile = root.openNextFile();
            if (gifFile)
            {
              if (!gifFile.isDirectory()) // play it
              {
                
                // C-strings... urghh...                
                memset(filePath, 0x0, sizeof(filePath));                
                strcpy(filePath, gifFile.path());
                
                // Show it.
                ShowGIF(filePath);
                gifFile.close();
              }


// Boot button Pattern advance with debounce
    int reading = digitalRead(BTN_PIN);
    if (reading != lastButtonState) {
      lastDebounceTime = millis();
    }
    
    if ((millis() - lastDebounceTime) > debounceDelay) {
      if (reading != buttonState) {
        buttonState = reading;
        if (buttonState == LOW) {
          Serial.println("NEXT GIF...");
           gifFile.close();
              gifFile = root.openNextFile();
            }
        root.close();
      } // root
        }
      }
    }
    
                
             
      
      delay(1000); // pause before restarting
      
   } // while
}`

**Comment by board707** [comment]
Hi... the formatting makes your code very hard to understand..

Copy your code to Arduino IDE, do autoalignment with Ctrl-T and you will see that your code follow that way::
1 open root
2 read 1st filename in the dir,  open and show GIF
3 test the button
4 if the button pressed - read the next filename
5 close root
6. return to 1 

Do you already see what is wrong? - when you return to stage 1 - you rewind the root and read the first file again.

You should think more about the logic of your program. 
And I don't think your problem is related to the library.

**Reply by GrimorJander** [reply]
> <img alt="" width="32" height="32" src="https://avatars.githubusercontent.com/u/50185434?s=64&amp;v=4">

I see what you are telling me, but I am quite a novice and it is a bit difficult for me. From what I see, I could eliminate step 5 (exit root) and thus keep it inside the GIF folder.
Or could I access root in void setup and get step 1 out of the loop? I never thought that the library was the problem. Just my incompetence. Thank you very much for dedicating your time to me.

**Comment by GrimorJander** [comment]
I'm still stuck in the loop. I only get to always play the same one or to pass all the gifs in order, but I can't  play one in a loop and change it to the next gif loop with just a button. I've even tried creating multiple folders for each image, or think of multiple buttons for each gif.
I know it's easier than all that, but I don't know how to do it.

**Comment by board707** [comment]
If you changed the code - show the new variant

**Reply by GrimorJander** [reply]
This is one with button working but only one GIF is playinig once when button is pushed.
No gif loop and no change with button as desired.
 Thanks in advance and sorry because english is not my native language. 


`// Example sketch which shows how to display a 64x32 animated GIF image stored in FLASH memory
// on a 64x32 LED matrix
//
// Credits: https://github.com/bitbank2/AnimatedGIF/tree/master/examples/ESP32_LEDMatrix_I2S
//

/* INSTRUCTIONS

   1. First Run the 'ESP32 Sketch Data Upload Tool' in Arduino from the 'Tools' Menu.
      - If you don't know what this is or see it as an option, then read this:
        https://github.com/me-no-dev/arduino-esp32fs-plugin
      - This tool will upload the contents of the data/ directory in the sketch folder onto
        the ESP32 itself.

   2. You can drop any animated GIF you want in there, but keep it to the resolution of the
      MATRIX you're displaying to. To resize a gif, use this online website: https://ezgif.com/

   3. Have fun.
*/

#define FILESYSTEM SPIFFS
#include <SPIFFS.h>
#include <AnimatedGIF.h>
#include <ESP32-HUB75-MatrixPanel-I2S-DMA.h>


// ----------------------------

// constants won't change. They're used here to set pin numbers:
const int buttonPin = 0;    // the number of the pushbutton pin
const int ledPin = 2;      // the number of the LED pin

// Variables will change:
int ledState = HIGH;         // the current state of the output pin
int buttonState;             // the current reading from the input pin
int lastButtonState = LOW;   // the previous reading from the input pin

// the following variables are unsigned longs because the time, measured in
// milliseconds, will quickly become a bigger number than can be stored in an int.
unsigned long lastDebounceTime = 0;  // the last time the output pin was toggled
unsigned long debounceDelay = 50;    // the debounce time; increase if the output flickers

// ----------------------------

/*
   Below is an is the 'legacy' way of initialising the MatrixPanel_I2S_DMA class.
   i.e. MATRIX_WIDTH and MATRIX_HEIGHT are modified by compile-time directives.
   By default the library assumes a single 64x32 pixel panel is connected.

   Refer to the example '2_PatternPlasma' on the new / correct way to setup this library
   for different resolutions / panel chain lengths within the sketch 'setup()'.

*/

#define PANEL_RES_X 64      // Number of pixels wide of each INDIVIDUAL panel module. 
#define PANEL_RES_Y 32     // Number of pixels tall of each INDIVIDUAL panel module.
#define PANEL_CHAIN 2      // Total number of panels chained one to another

//MatrixPanel_I2S_DMA dma_display;
MatrixPanel_I2S_DMA *dma_display = nullptr;

uint16_t myBLACK = dma_display->color565(0, 0, 0);
uint16_t myWHITE = dma_display->color565(255, 255, 255);
uint16_t myRED = dma_display->color565(255, 0, 0);
uint16_t myGREEN = dma_display->color565(0, 255, 0);
uint16_t myBLUE = dma_display->color565(0, 0, 255);


AnimatedGIF gif;
File f;
int x_offset, y_offset;



// Draw a line of image directly on the LED Matrix
void GIFDraw(GIFDRAW *pDraw)
{
  uint8_t *s;
  uint16_t *d, *usPalette, usTemp[320];
  int x, y, iWidth;

  iWidth = pDraw->iWidth;
  if (iWidth > MATRIX_WIDTH)
    iWidth = MATRIX_WIDTH;

  usPalette = pDraw->pPalette;
  y = pDraw->iY + pDraw->y; // current line

  s = pDraw->pPixels;
  if (pDraw->ucDisposalMethod == 2) // restore to background color
  {
    for (x = 0; x < iWidth; x++)
    {
      if (s[x] == pDraw->ucTransparent)
        s[x] = pDraw->ucBackground;
    }
    pDraw->ucHasTransparency = 0;
  }
  // Apply the new pixels to the main image
  if (pDraw->ucHasTransparency) // if transparency used
  {
    uint8_t *pEnd, c, ucTransparent = pDraw->ucTransparent;
    int x, iCount;
    pEnd = s + pDraw->iWidth;
    x = 0;
    iCount = 0; // count non-transparent pixels
    while (x < pDraw->iWidth)
    {
      c = ucTransparent - 1;
      d = usTemp;
      while (c != ucTransparent && s < pEnd)
      {
        c = *s++;
        if (c == ucTransparent) // done, stop
        {
          s--; // back up to treat it like transparent
        }
        else // opaque
        {
          *d++ = usPalette[c];
          iCount++;
        }
      } // while looking for opaque pixels
      if (iCount) // any opaque pixels?
      {
        for (int xOffset = 0; xOffset < iCount; xOffset++ ) {
          dma_display->drawPixel(x + xOffset, y, usTemp[xOffset]); // 565 Color Format
        }
        x += iCount;
        iCount = 0;
      }
      // no, look for a run of transparent pixels
      c = ucTransparent;
      while (c == ucTransparent && s < pEnd)
      {
        c = *s++;
        if (c == ucTransparent)
          iCount++;
        else
          s--;
      }
      if (iCount)
      {
        x += iCount; // skip these
        iCount = 0;
      }
    }
  }
  else // does not have transparency
  {
    s = pDraw->pPixels;
    // Translate the 8-bit pixels through the RGB565 palette (already byte reversed)
    for (x = 0; x < pDraw->iWidth; x++)
    {
      dma_display->drawPixel(x, y, usPalette[*s++]); // color 565
    }
  }
} /* GIFDraw() */


void * GIFOpenFile(const char *fname, int32_t *pSize)
{
  Serial.print("Playing gif: ");
  Serial.println(fname);
  f = FILESYSTEM.open(fname);
  if (f)
  {
    *pSize = f.size();
    return (void *)&f;
  }
  return NULL;
} /* GIFOpenFile() */

void GIFCloseFile(void *pHandle)
{
  File *f = static_cast<File *>(pHandle);
  if (f != NULL)
    f->close();
} /* GIFCloseFile() */

int32_t GIFReadFile(GIFFILE *pFile, uint8_t *pBuf, int32_t iLen)
{
  int32_t iBytesRead;
  iBytesRead = iLen;
  File *f = static_cast<File *>(pFile->fHandle);
  // Note: If you read a file all the way to the last byte, seek() stops working
  if ((pFile->iSize - pFile->iPos) < iLen)
    iBytesRead = pFile->iSize - pFile->iPos - 1; // <-- ugly work-around
  if (iBytesRead <= 0)
    return 0;
  iBytesRead = (int32_t)f->read(pBuf, iBytesRead);
  pFile->iPos = f->position();
  return iBytesRead;
} /* GIFReadFile() */

int32_t GIFSeekFile(GIFFILE *pFile, int32_t iPosition)
{
  int i = micros();
  File *f = static_cast<File *>(pFile->fHandle);
  f->seek(iPosition);
  pFile->iPos = (int32_t)f->position();
  i = micros() - i;
  //  Serial.printf("Seek time = %d us\n", i);
  return pFile->iPos;
} /* GIFSeekFile() */

unsigned long start_tick = 0;

void ShowGIF(char *name)
{
  start_tick = millis();

  if (gif.open(name, GIFOpenFile, GIFCloseFile, GIFReadFile, GIFSeekFile, GIFDraw))
  {
    x_offset = (MATRIX_WIDTH - gif.getCanvasWidth()) / 2;
    if (x_offset < 0) x_offset = 0;
    y_offset = (MATRIX_HEIGHT - gif.getCanvasHeight()) / 2;
    if (y_offset < 0) y_offset = 0;
    Serial.printf("Successfully opened GIF; Canvas size = %d x %d\n", gif.getCanvasWidth(), gif.getCanvasHeight());
    Serial.flush();
    while (gif.playFrame(true, NULL))
    {
      if ( (millis() - start_tick) > 8000) { // we'll get bored after about 8 seconds of the same looping gif
        break;
      }
    }
    gif.close();
  }

} /* ShowGIF() */



/************************* Arduino Sketch Setup and Loop() *******************************/
void setup() {



  //
  pinMode(buttonPin, INPUT);
  pinMode(ledPin, OUTPUT);

  // set initial LED state
  digitalWrite(ledPin, ledState);
  //

  HUB75_I2S_CFG mxconfig(
    PANEL_RES_X,   // module width
    PANEL_RES_Y,   // module height
    PANEL_CHAIN    // Chain length
  );

  mxconfig.gpio.e = 18;
  mxconfig.clkphase = false;
  //mxconfig.driver = HUB75_I2S_CFG::FM6126A;
  //mxconfig.driver = HUB75_I2S_CFG::SHIFT;   // shift reg driver, default is plain shift register
  //mxconfig.double_buff = false;             // use double buffer (twice amount of RAM required)
  //mxconfig.i2sspeed = HUB75_I2S_CFG::HZ_10M;// I2S clock speed, better leave as-is unless you want to experiment




  // Display Setup
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);
  dma_display->begin();
  dma_display->setBrightness8(90); //0-255
  dma_display->clearScreen();
  dma_display->fillScreen(myWHITE);

  //


  Serial.begin(115200);
  Serial.println("Starting AnimatedGIFs Sketch");

  // Start filesystem
  Serial.println(" * Loading SPIFFS");
  if (!SPIFFS.begin()) {
    Serial.println("SPIFFS Mount Failed");
  }

  dma_display->begin();

  /* all other pixel drawing functions can only be called after .begin() */
  dma_display->fillScreen(dma_display->color565(0, 0, 0));
  gif.begin(LITTLE_ENDIAN_PIXELS);

}

String gifDir = "/gifs"; // play all GIFs in this directory on the SD card
char filePath[256] = { 0 };
File root, gifFile;

void loop()
{
  root = FILESYSTEM.open(gifDir);
  // read the state of the switch into a local variable:
  int reading = digitalRead(buttonPin);

  // check to see if you just pressed the button
  // (i.e. the input went from LOW to HIGH), and you've waited long enough
  // since the last press to ignore any noise:

  // If the switch changed, due to noise or pressing:
  if (reading != lastButtonState) {
    // reset the debouncing timer
    lastDebounceTime = millis();
  }

  if ((millis() - lastDebounceTime) > debounceDelay) {
    // whatever the reading is at, it's been there for longer than the debounce
    // delay, so take it as the actual current state:

    // if the button state has changed:
    if (reading != buttonState) {
      buttonState = reading;

      // only toggle the LED if the new button state is HIGH
      if (buttonState == HIGH) {
        ledState = !ledState;
        Serial.println("Button push Detected");

        gifFile = root.openNextFile();
        while (gifFile)
        {
          if (!gifFile.isDirectory()) // play it
          {

            // C-strings... urghh...
            memset(filePath, 0x0, sizeof(filePath));
            strcpy(filePath, gifFile.path());

            // Show it.
            ShowGIF(filePath);

          }
          gifFile.close();
          //gifFile = root.openNextFile();
        }
        root.close();
      } // root
    }
  }
  // set the LED:
  digitalWrite(ledPin, ledState);

  // save the reading. Next time through the loop, it'll be the lastButtonState:
  lastButtonState = reading;


}


`
[AnimatedGifProbando2.zip](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/files/9101652/AnimatedGifProbando2.zip)

**Reply by GrimorJander** [reply]
Another approach could be to create as many folders as gif images I want to loop (gifDir1, gifDir2, gifDir3...), and then in the code tell the button to change folders with the command:   root = FILESYSTEM.open(gifDir) .
But then gifDir should be incremental (gifDir++ ?)

**Comment by ergindemir** [comment]
I would first assign internal pull up resistor to the button for stable operation.

**Reply by GrimorJander** [reply]
I can try it, although my skills are very limited. One thing I've noticed with the button in this sketch is that it only recognizes the press after the entire gif playback sequence. As if it had a very long delay...

**Reply by board707** [reply]
@GrimorJander 
Oh, I somehow didn’t think that you might have problems with the button.
And if you temporarily remove the button completely (as the test) - will your code sequentially show all the files in the directory?

**Reply by GrimorJander** [reply]
I can see the secuence only in serial monitor. Rgb display matrix doesn't show anything with my modifications.

**Reply by board707** [reply]
But if you run AnumatedGIF example unchanged - it works?

I think you should advancing sequentially. 
First - make sure that the standard example works for your hardware setup. 
Second - change one thing (for example, add showing next image from the dir after current) - and test how it works. If there are any problems - fix it before going forward.
Next - add playing all images in the loop... fix errors...
Next - add the button...

and so on... 

Without going through each step, there is little chance of solving the problem. As I see,  all what you have now - the complicated mixture of broken code... nothing works as expected...

**Reply by GrimorJander** [reply]
> I would first assign internal pull up resistor to the button for stable operation.

Thanks , Pulling Up internally seens to improve the performance.

> Without going through each step, there is little chance of solving the problem. As I see, all what you have now - the complicated mixture of broken code... nothing works as expected...

You´recompletly righ, i´ll try to clean the code without break it.

Now my skecht seens to work with both, button interaction and displaying gifs on the matrix.
Here´s my loop righ now (showing only first gif, one time, and everytime i press button) Yeah!!

`void loop()
{
  root = FILESYSTEM.open(gifDir);
  buttonState = digitalRead(buttonPin);

  // check if the pushbutton is pressed. If it is, the buttonState is LOW:
  if (buttonState == LOW) {
    // turn LED on:
    Serial.println("Button Press");
    gifFile = root.openNextFile();



    // C-strings... urghh...
    memset(filePath, 0x0, sizeof(filePath));
    strcpy(filePath, gifFile.path());

    // Show it.
    ShowGIF(filePath);

  }
}
`
So, How can i manage to loop first gif and when press button change the loop to the next gif?

**Reply by board707** [reply]
With ShowGIF() function duration about 8 sec and your button non-operation during all this time it hardly can be stated useful.  I think you should add checking the button state inside the ShowGIF()

**Reply by GrimorJander** [reply]
> With ShowGIF() function duration about 8 sec and your button non-operation during all this time it hardly can be stated useful. I think you should add checking the button state inside the ShowGIF()

Well, button only respond when the loop restart, but not going to press it so often, and single gif duration is about 3 seconds or less.
The final project is a letter of 128x32 matrix in the door of my officce, where usually make live streams for news-tv. So i will need says 5 different animated  lettering or simbols (Welcome - Name of the company - On Air - Do Not Enter - Breakfast Break) .
I could change it from pc with "Skecht Data Upload" method, but the rest of the people in the officce dont know how to do it so the button is preffered.

**Reply by board707** [reply]
> The final project is a letter of 128x32 matrix in the door of my officce, where usually make live streams for news-tv. So i will need says 5 different animated lettering or simbols (Welcome - Name of the company - On Air - Do Not Enter - Breakfast Break) .

It looks very interesting, please share the links when the project will be ready. Maybe I'll do something similar on stm32 boards :)

**Reply by GrimorJander** [reply]
> It looks very interesting, please share the links when the project will be ready. Maybe I'll do something similar on stm32 boards :)

We are currently running the scrolling text with Telegram Bot from Brian Lough: https://github.com/witnessmenow/ScrollingLEDMatrixTelegram
But´s it´s way more visual, modern and funny with animated gifs. I know for sure how to make animated messages from video and then video to gif, but not how to code C++.

![photo](https://user-images.githubusercontent.com/27100419/179197312-9dc9faff-5bd3-47ef-9d58-c27e5c442e5a.jpg)

**Reply by board707** [reply]
> ![photo](https://user-images.githubusercontent.com/27100419/179197312-9dc9faff-5bd3-47ef-9d58-c27e5c442e5a.jpg)

great ^)

> I know for sure how to make animated messages from video and then video to gif, but not how to code C++.

opposite for me, I don't understand anything in the video and gifs

**Reply by board707** [reply]
@GrimorJander 
Is it possible to make animated GIF from this picture? :) (it is RGB 64x64 matrix)

![P_20220222_111258small](https://user-images.githubusercontent.com/50185434/179369608-1bc0d0e3-df09-4cf4-b815-c41d26b2d753.jpg)

**Reply by GrimorJander** [reply]
It seems to be a clock with custom pixel flower frame. Probably the flower is static and drawn by pixel mapping.
So much more difficult will be join clock with animated gif skecht.

**Reply by board707** [reply]
> So much more difficult will be join clock with animated gif skecht.

It seems not so difficult to join it together in the matrix if dual buffering is used. As I think... first update the next frame of animation on whole panel,  next delete rectangle in the centre and after that draw the clock over the picture. Using of dual buffering prevent displaying picture artefacts.

**Comment by ergindemir** [comment]
Good, I'm happy the 1st step worked well.
2nd step: I would pick 5 images and save them as 0.gif, 1.gif, .., 4.gif
Define a global variable int i, increment and take modulo 5 at each button press and show i.gif .
If you make this work successfully you can extend your code to arbitrary files read from directory.

**Reply by GrimorJander** [reply]
> Good, I'm happy the 1st step worked well. 2nd step: I would pick 5 images and save them as 0.gif, 1.gif, .., 4.gif Define a global variable int i, increment and take modulo 5 at each button press and show i.gif . If you make this work successfully you can extend your code to arbitrary files read from directory.

Images are already renamed this way on Gifs directory. Do you have an example skecht of global variable + increment + modulo 5 thing?
I can only read and cath pieces of code and make a frankenstein code, but not from scratch

**Reply by board707** [reply]
> do you have an example skecht of global variable + increment + modulo 5 thing?

increment + modulo 5:

(++i)%5

**Reply by ergindemir** [reply]
i++;
i%=5;
printf("%d.gif",i);

**Reply by GrimorJander** [reply]
> "%d.gif",i

I don´t know how implement this in my loop. 
compiling error says: 'i' was not declared in this scope

**Comment by davemaster** [comment]
cmon, You get HERE, and have problems with a button?, try one example, using button first, then.... yadi yadi ...........

**Reply by GrimorJander** [reply]
Yes, the Button of my dead.

I add a  `int count = 0` 
 And my loop look this way now:

` 
void loop() 
{
root = FILESYSTEM.open(gifDir);
  buttonState = digitalRead(buttonPin);

  // check if the pushbutton is pressed. If it is, the buttonState is HIGH:
  if (buttonState == LOW) {
   
    count++;
        Serial.println(count);
    gifFile = root.openNextFile();
    
         
           
                
                // C-strings... urghh...                
                memset(filePath, 0x0, sizeof(filePath));                
                strcpy(filePath, gifFile.path());
                
                // Show it.
                ShowGIF(filePath);
               
              }
              }
              
 
      `

So i can see now a incremental counter on serial monitor... but , keep playin same gif.
So how to say in C++ to show next incremental gif to the loop routine.

**Comment by davemaster** [comment]
do a "list" of all gif files, then point each file as You wish

El dom, 17 jul 2022 a las 7:53, GrimorJander ***@***.***>)
escribió:

> Yes, the Button of my dead.
>
> I add a int count = 0
> And my loop look this way now:
>
> `void loop()
> {
> root = FILESYSTEM.open(gifDir);
> buttonState = digitalRead(buttonPin);
>
> // check if the pushbutton is pressed. If it is, the buttonState is HIGH:
> if (buttonState == LOW) {
>
> count++;
>     Serial.println(count);
> gifFile = root.openNextFile();
>
>
>
>
>             // C-strings... urghh...
>             memset(filePath, 0x0, sizeof(filePath));
>             strcpy(filePath, gifFile.path());
>
>             // Show it.
>             ShowGIF(filePath);
>
>           }
>           }
>
>
>   `
>
> So i can see now a incremental counter on serial monitor... but , how to
> say in C++ to show next incremental gif to the loop routine.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/discussions/297#discussioncomment-3165093>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIGUX5CBJADCFGNVVFUUSDVUP63VANCNFSM523H6JJA>
> .
> You are receiving this because you commented.Message ID:
> <mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/repo-discussions/297/comments/3165093
> @github.com>
>


-- 
David Elias Flores Escalante
TeleTracking SAC

**Reply by GrimorJander** [reply]
Hola, David. 
It's difficult to me to point to each file because I only know 2 commands of the SPIFFS funtions: openNextfile and openDir but when I try openfile ("path", "r") nothing happen

**Reply by board707** [reply]
> It's difficult to me to point to each file because I only know 2 commands of the SPIFFS funtions:

we all do not understand something at first, but then we can learn.

**Comment by davemaster** [comment]
then search for SPIFFS examples...

**Comment by GrimorJander** [comment]
Finally got it working!!
I have made my own logical sequence (totally contrary to the principles of good programming), but the fact is that it works.
The issue was that I couldn't get the directory initialization out of the void loop because it gave a compile error. So it occurred to me to make it run only once while inside the loop by associating it with the cycle counter. So it enters the directory only once in the first cycle and then it already repeat gif in loop or change with the button. Mission achieved.
The only thing left to fix is ​​that when you reach the last file when you press next the program crashes and restarts. It's not a big deal because it's very fast and it reloads the first gif, which is what it's all about, but it would be more elegant if I could get back to the start if there's nothing else to show.
I want to thank you all for the help and advice.
I put the loop part again:

`void loop()
{
  if (count < 2) {
    root = FILESYSTEM.open(gifDir);
    gifFile = root.openNextFile();
  }

  buttonState = digitalRead(buttonPin);

  // check if the pushbutton is pressed. If it is, the buttonState is HIGH:
  if (buttonState == LOW) {

    Serial.println("Button Press");
    gifFile = root.openNextFile();



    // C-strings... urghh...
    memset(filePath, 0x0, sizeof(filePath));
    strcpy(filePath, gifFile.path());

    // Show it.
    ShowGIF(filePath);

  }
  if (buttonState == HIGH) {

    // C-strings... urghh...
    memset(filePath, 0x0, sizeof(filePath));
    strcpy(filePath, gifFile.path());
    ShowGIF(filePath);
    count++;
    Serial.println(count);
  }


}
`

**Comment by vortigont** [comment]
@GrimorJander, maybe too late but you might wanna check this lib [InterruptButton](https://github.com/rwmingis/InterruptButton). With that lib you do not need any loop() tricks, it works on interrupts out of loop() scope. Makes life much easier.
